domain/secret/state/state.go

```diff
@@ -77,19 +79,181 @@ func (st State) getModelUUID(ctx context.Context, tx *sqlair.TX) (string, error)
+		return false, nil
+	}
+
+	input := secretApplicationOwner{Label: label, ApplicationUUID: appUUID.String()}
+	count := Count{}
+	// TODO(secrets) - we check using 2 queries, but should do in DDL
+	checkLabelExistsSQL := `
+SELECT COUNT(*) AS &Count.num
+FROM (
+    SELECT secret_id
+    FROM   secret_application_owner
+    WHERE  label = $secretApplicationOwner.label
+    AND    application_uuid = $secretApplicationOwner.application_uuid
+    UNION
+    SELECT secret_id
+    FROM   secret_unit_owner
+       JOIN unit u ON u.uuid = unit_uuid
+    WHERE  label = $secretApplicationOwner.label
+    AND    u.application_uuid = $secretApplicationOwner.application_uuid
+)`
+
+	checkExistsStmt, err := st.Prepare(checkLabelExistsSQL, input, count)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	err = domain.Run(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, checkExistsStmt, input).Get(&count)
+		return errors.Trace(err)
+	})
+	if err != nil {
+		return false, fmt.Errorf("checking if secret owned by application %q with label %q exists: %w", appUUID, label, err)
+	}
+	return count.Num > 0, nil
+}
+
+// CheckUnitSecretLabelExists returns function which checks if a charm unit secret with the given label already exists.
+func (st State) CheckUnitSecretLabelExists(ctx domain.AtomicContext, unitUUID coreunit.UUID, label string) (bool, error) {
+	if label == "" {
+		return false, nil
+	}
+
+	input := secretUnitOwner{Label: label, UnitUUID: unitUUID.String()}
+	count := Count{}
+	// TODO(secrets) - we check using 2 queries, but should do in DDL
+	checkLabelExistsSQL := `
+SELECT COUNT(*) AS &Count.num
+FROM (
+    SELECT secret_id
+    FROM   secret_application_owner sao
+           JOIN unit u ON sao.application_uuid = u.application_uuid
+    WHERE  label = $secretUnitOwner.label
+    AND    u.uuid = $secretUnitOwner.unit_uuid
+    UNION
+    SELECT DISTINCT secret_id
+    FROM   secret_unit_owner suo
+           JOIN unit u ON suo.unit_uuid = u.uuid
+           JOIN unit peer ON peer.application_uuid = u.application_uuid
+    WHERE  label = $secretUnitOwner.label
+    AND peer.uuid != u.uuid
+)`
+
+	checkExistsStmt, err := st.Prepare(checkLabelExistsSQL, input, count)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	err = domain.Run(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, checkExistsStmt, input).Get(&count)
+		return errors.Trace(err)
+	})
+	if err != nil {
+		return false, fmt.Errorf(
+			"checking if secret owned by unit %q with label %q exists: %w", unitUUID, label, err)
+	}
+	return count.Num > 0, nil
+
+}
+
+// CheckUserSecretLabelExists returns an error if a user secret with the given label already exists.
+func (st State) CheckUserSecretLabelExists(ctx domain.AtomicContext, label string) (bool, error) {
+	if label == "" {
+		return false, nil
+	}
+	input := secretOwner{Label: label}
+	count := Count{}
+	checkLabelExistsSQL := `
+SELECT COUNT(*) AS &Count.num
```

Comment:
We don't need Count{}
Just `select true` and check for `sqlair.NoRows`
No need for an extra type struct